bundles/org.openhab.binding.linuxinput/src/main/java/org/openhab/binding/linuxinput/internal/DeviceReadingHandler.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.linuxinput.internal;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.CancellationException;
+
+/**
+ * Abstract handler, that encapsulates the lifecycle of an underlying device.
+ *
+ * @author Thomas Weißschuh - Initial contribution
+ */
+@NonNullByDefault
+public abstract class DeviceReadingHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(DeviceReadingHandler.class);
+
+    private @NonNullByDefault({}) Thread worker = null;
+
+    abstract boolean immediateSetup() throws IOException;
+
+    abstract boolean delayedSetup() throws IOException;
+
+    abstract void handleEventsInThread() throws IOException;
+
+    abstract void closeDevice() throws IOException;
+
+    abstract String getInstanceName();
+
+    public DeviceReadingHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public final void initialize() {
+        boolean performDelayedSetup = performImmediateSetup();
+        if (performDelayedSetup) {
+            scheduler.execute(() -> {
+                boolean handleEvents = performDelayedSetup();
+                if (handleEvents) {
+                    Thread thread = Utils.backgroundThread(() -> {
+                        try {
+                            handleEventsInThread();
+                        } catch (IOException e) {
+                            logger.warn("Could not read event", e);
+                            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                        }
+                    } , getClass(), getInstanceName());
+                    thread.start();
+                    worker = thread;
+                }
+            });
+        }
+    }
+
+    private boolean performImmediateSetup() {
+        try {
+            return immediateSetup();
+        } catch (IOException e) {
+            handleSetupError(e);
+            return false;
+        }
+    }
+
+    private boolean performDelayedSetup() {
+        try {
+            return delayedSetup();
+        } catch (IOException e) {
+            handleSetupError(e);
+            return false;
+        }
+    }
+
+    private void handleSetupError(Exception e) {
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+    }
+
+    @Override
+    public final void dispose() {
```

Comment:
It should be instant.
What about reducing the timeout to 100ms or so?
I would like to make sure everything is cleaned up when `dispose()` returns instead of putting it into a background thread.

Reply:
> What about reducing the timeout to 100ms or so?

Yes that would be great 👍